Prepare ShardFollowNodeTask to bootstrap when it fall behind leader shard

martijnvg opened the pull request:
* Changed the shard changes api to include a special metadata in the exception being thrown
to indicate that the ops are no longer there.
* Changed ShardFollowNodeTask to handle this exception with special metadata
  and mark a shard as fallen behind its leader shard. The shard follow task
  will then abort its on going replication.

The code that does the restore from ccr repository still needs to be added.
This change should make that change a bit easier.

Relates to #35975

elasticmachine commented:
Pinging @elastic/es-distributed

dnhatn commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java:
```diff
@@ -397,7 +398,11 @@ protected void doExecute(Task task, Request request, ActionListener<Response> li
                     String message = "Operations are no longer available for replicating. Maybe increase the retention setting [" +
                         IndexSettings.INDEX_SOFT_DELETES_RETENTION_OPERATIONS_SETTING.getKey() + "]?";
-                    listener.onFailure(new ElasticsearchException(message, e));
+                    // Make it easy to detect this error in ShardFollowNodeTask:
```
Maybe introduce a new exception (extends from IllegalStateException) which indicates that the requesting history is no longer available. We can throw that exception in the LuceneChangesSnapshot. WDYT?

dnhatn replied:
Thanks @martijnvg :)

martijnvg commented:
@dnhatn I've updated the PR.

dnhatn commented on server/src/main/java/org/elasticsearch/index/engine/OperationsMissingException.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public final class OperationsMissingException extends IllegalStateException {
+
+    OperationsMissingException(String message) {
```
Maybe name it `MissingHistoryOperationsException`?

dnhatn commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java:
```diff
@@ -274,6 +277,13 @@ private void sendShardChangesRequest(long from, int maxOperationCount, long maxR
                         failedReadRequests++;
                         fetchExceptions.put(from, Tuple.tuple(retryCounter, ExceptionsHelper.convertToElastic(e)));
                     }
+                    if (e instanceof ElasticsearchException) {
```
Can we serialize the newly added exception and use it here?

martijnvg replied:
But then `MissingHistoryOperationsException` needs to extend `ElasticsearchException`, otherwise it cannot be serialized.

Regardless of that I always understood that we want to minimize the number of exceptions that are serializable. This is why I added a header to `ElasticsearchException` instead of creating a new one. Does this still apply? Or do we really want to intoduce a new serializable exception here?

dnhatn replied:
> Regardless of that I always understood that we want to minimize the number of exceptions that are serializable. 

If this is the case, let's use the header or metadata here.

dnhatn commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java:
```diff
@@ -290,6 +300,17 @@ void handleReadResponse(long from, long maxRequiredSeqNo, ShardChangesAction.Res
     }
 
+    void handleFallenBehindLeaderShard(Exception e) {
+        if (fallenBehindLeaderShard.compareAndSet(false, true)) {
```
Let's make this method noop for now (i.e., without the state `fallenBehindLeaderShard`). I feel we need a more robust approach to avoid the scenario where an outstanding request can trigger another restore while the shard was restored already.

martijnvg commented:
@dnhatn I've updated the PR.

martijnvg commented:
run the gradle build tests 1

martijnvg commented:
@elasticmachine run elasticsearch-ci/2

martijnvg commented:
@elasticmachine run elasticsearch-ci/1

martijnvg commented:
@dnhatn As discussed, I have removed the `fallen_behind_leader_shard` stats field and `fallenBehindLeaderShard` field, because it is unknown whether a simple flag is sufficient enough
to indicate we need to fallback to a file based restore.

tbrooks8 commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java:
```diff
@@ -274,6 +275,13 @@ private void sendShardChangesRequest(long from, int maxOperationCount, long maxR
                     }
+                    if (e instanceof ElasticsearchException) {
+                        ElasticsearchException elasticsearchException = (ElasticsearchException) e;
+                        if (elasticsearchException.getMetadataKeys().contains(Ccr.REQUESTED_OPS_MISSING_METADATA_KEY)) {
```
I have been working on this PR locally with some other work of mine. When the listener is called, the exception is wrapped in a `RemoteTransportException`. So the check if the metadata is present fails even if the underlying exception has the metadata. I think that this needs to be unwrapped further. Blocking on a future unwraps the exception which is why your test passes (opposed to the listener infrastructure).

martijnvg replied:
I think also because the `MissingHistoryOperationsException` was not handeled properly in shard changes action, the `ElasticsearchException` with the metadata never made it to here.

tbrooks8 commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java:
```diff
@@ -394,10 +396,14 @@ protected void asyncShardOperation(
             ActionListener<Response> wrappedListener = ActionListener.wrap(listener::onResponse, e -> {
                 Throwable cause = ExceptionsHelper.unwrapCause(e);
-                if (cause instanceof IllegalStateException && cause.getMessage().contains("Not all operations between from_seqno [")) {
+                if (cause instanceof MissingHistoryOperationsException) {
```
I don't think this is serializing properly. I find this exception at a breakpoint here. I think that our exception serializing code is doing an `instanceof` check, seeing that this is an `IllegalStateException` and it is identified as an `IllegalStateException` on deserialization.

```
java.lang.IllegalStateException: Not all operations between from_seqno [17] and to_seqno [34] found; expected seqno [17]; found [Index{id='6', type='doc', seqNo=18, primaryTerm=1, version=2, autoGeneratedIdTimestamp=-1}]
	at org.elasticsearch.index.engine.LuceneChangesSnapshot.rangeCheck(LuceneChangesSnapshot.java:155)
	at org.elasticsearch.index.engine.LuceneChangesSnapshot.next(LuceneChangesSnapshot.java:138)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction.getOperations(ShardChangesAction.java:527)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction$TransportAction.shardOperation(ShardChangesAction.java:340)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction$TransportAction.shardOperation(ShardChangesAction.java:319)
	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$1.doRun(TransportSingleShardAction.java:117)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:732)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
```

martijnvg replied:
I made a wrong assumption here. I thought that the exception here didn't get serialized, but it does get serialized. (when the request gets send to a different shard copy then the local node has).

This means that throwing the `MissingHistoryOperationsException` here does not work. We need to throw an exception that extends from `ElasticticsearchException` and that sets specific metadata in the LuceneSnapshot.java . I will make this change now.

martijnvg commented:
@dnhatn @tbrooks8 I've updated the PR. I removed the `MissingHistoryOperationsException` and instead throw now `ResourceNotFoundException` with headers from `LuceneSnapshot#rangeCheck(...)`. This is needed, because`MissingHistoryOperationsException` was being serialized and then it gets wrapped in `RemoteTransportException` and then we can't handle it properly.

@tbrooks8 Can you verify whether this change works in your PR?

martijnvg commented:
@dnhatn and I talked about keeping the `MissingHistoryOperationsException` the was removed. We think we can still use it, but we need to handle the exception in a different place. 

So I reverted my previous commit and instead of handeling the `MissingHistoryOperationsException` in `doExecute(...)` method, the `MissingHistoryOperationsException` is handled in `getOperations(...)` which is always executed locally on the node that has the shard copy.

The only minor downside of this approach is that will potentially wrap a `MissingHistoryOperationsException` more than once. A shard copy may not have the requested ops while another shard copy does have the requested ops. In the end that shouldn't be a problem because `TransportSingleShardAction` always keep track of the latest failure.

tbrooks8 commented:
I reran my test and found the `handleFallenBehindLeaderShard` was called this time.
